### ISSUE TITLE
Strange training assert

### ISSUE THREAD

**Issue by sineptic** [open-spaced-repetition/fsrs-rs#233]
When I want to compute params why it is necessary to have item WITH EXACTLY ONE `long term review` and if it doesn't exist(or all items contain more) you library return `NotEnoughData`???

to reproduce this you must have more than 7 items, because otherwise `DEFAULT_PARAMETERS` returned.

items partitioned at the dataset.rs:236

**Comment by L-M-Sherlock** [comment]
see: https://github.com/open-spaced-repetition/fsrs-rs/pull/172

**Comment by sineptic** [comment]
but your commit bot change it i think

**Comment by L-M-Sherlock** [comment]
> why it is necessary to have item WITH EXACTLY ONE long term review and if it doesn't exist(or all items contain more) you library return NotEnoughData

Because only long-term reviews' results are treated as the labels during optimization.

**Comment by sineptic** [comment]
but why == 1 , not >= 1?

**Comment by L-M-Sherlock** [comment]
> but why == 1 , not >= 1?

Because the pretrainset only need items whose long term reviews ==1.

**Comment by sineptic** [comment]
Maybe I don't understand smth. What is `pretrainset`?
Or what I must do to compute parameters?

**Comment by L-M-Sherlock** [comment]
FSRS-5 has 19 parameters. The first four parameters are the initial stability, which is optimized via fitting forgetting curves in the pretrainset, where the data only have one long-term reviews.

**Comment by sineptic** [comment]
Ok, but isnt this car amount small? why not use just first result of all?

**Comment by L-M-Sherlock** [comment]
The error is raised here:

https://github.com/open-spaced-repetition/fsrs-rs/blob/189501e3e6133e2d57ced346b4aa5ea6685edea2/src/pre_training.rs#L203-L205

So I change my previous reasoning. Your pretrainset doesn't have data with one long-term review.

**Comment by sineptic** [comment]
https://github.com/open-spaced-repetition/fsrs-rs/blob/189501e3e6133e2d57ced346b4aa5ea6685edea2/src/dataset.rs#L234-L236

this split items, and do it in strange way. Why you can't use other cards?

**Comment by sineptic** [comment]
At least!!! but you code check exactly 1.

``` rust
use fsrs::{FSRSItem, FSRSReview, DEFAULT_PARAMETERS, FSRS};

fn main() {
    let fsrs = FSRS::new(Some(&DEFAULT_PARAMETERS)).unwrap();

    let item = FSRSItem {
        reviews: vec![
            FSRSReview {
                rating: 1,
                delta_t: 0,
            },
            FSRSReview {
                rating: 1,
                delta_t: 1,
            },
            FSRSReview {
                rating: 1,
                delta_t: 1,
            },
        ],
    };
    let items = vec![item; 140];

    let _best_params = fsrs.compute_parameters(items, None).unwrap();
}
```
This return error

**Comment by L-M-Sherlock** [comment]
This returns error because the pretrainset is still empty. All items have two long-term reviews and none of them include one long-term review.

**Comment by sineptic** [comment]
I dont understand how it's related. 
Also tests require `collection.anki21` file.

**Comment by sineptic** [comment]
I read whole description of algorithm, but don't found pretrainset.
I now can't understand why we can't make item with 1 long-term repetition by removing second.

**Comment by L-M-Sherlock** [comment]
> Also tests require `collection.anki21` file.

You can remove `anki21_sample_file_converted_to_fsrs()` from the list to only run the test in csv.

https://github.com/open-spaced-repetition/fsrs-rs/blob/189501e3e6133e2d57ced346b4aa5ea6685edea2/src/training.rs#L479

> I read whole description of algorithm, but don't found pretrainset.

Algorithm only include the information about how the algorithm works. It doesn't explain how to optimize it.

> I currently can't understand why we can't make item with 1 long-term repetition by removing second.

What do you mean by "removing second"?

**Comment by sineptic** [comment]
> Algorithm only include the information about how the algorithm works. It doesn't explain how to optimize it.

So I don't understand what is pretrainset.


> What do you mean by "removing second"?
``` rust
vec![
    FSRSReview {
        rating: 1,
        delta_t: 0,
    },
    FSRSReview {
        rating: 1,
        delta_t: 1,
    },
    FSRSReview {
        rating: 1,
        delta_t: 1,
    },
]

// Treat as

vec![
    FSRSReview {
        rating: 1,
        delta_t: 0,
    },
    FSRSReview { // First long-term repetition
        rating: 1,
        delta_t: 1,
    },
    /*
        tail...
    */
]
```

**Comment by L-M-Sherlock** [comment]
> So I don't understand what is pretrainset.

The `pretrainset` is used to optimized the first four parameters of FSRS, which are related to the four initial stabilities corresponding to four ratings of the first review.

**Comment by L-M-Sherlock** [comment]
> ```rust
> vec![
>     FSRSReview {
>         rating: 1,
>         delta_t: 0,
>     },
>     FSRSReview {
>         rating: 1,
>         delta_t: 1,
>     },
>     FSRSReview {
>         rating: 1,
>         delta_t: 1,
>     },
> ]
> 
> // Treat as
> 
> vec![
>     FSRSReview {
>         rating: 1,
>         delta_t: 0,
>     },
>     FSRSReview { // First long-term repetition
>         rating: 1,
>         delta_t: 1,
>     },
>     /*
>         tail...
>     */
> ]
> ```

The function treats every `vec` as one sample of the dataset. So you need to make a copy of the original vec and then remove the second reviews by yourself.

**Comment by sineptic** [comment]
but why library doesn't do this?

**Comment by L-M-Sherlock** [comment]
> but why library doesn't do this?

Because this library is only used by Anki. These works have been done in Anki side. So I don't have motivation to implement this in the library.

**Comment by sineptic** [comment]
OK, thank you for work.

**Comment by sineptic** [comment]
> These works have been done in Anki side. 

Is it true, that when I add this (1 long-term repetition) items, other(not first 4 parameters) stay the same?

**Comment by L-M-Sherlock** [comment]
> Is it true, that when I add this (1 long-term repetition) items, other(not first 4 parameters) stay the same?

Yes, because the optimization are divided into two parts: pre-training and training. If you only provide items with one long-term reviewed, the training part will be skipped.

**Comment by sineptic** [comment]
Is there any other things, that implemented on Anki side to use this library, or this is only issue?

**Comment by L-M-Sherlock** [comment]
If I remember correctly, they have been implemented in convertor_test.rs for test use. You can refer the code.

**Comment by sineptic** [comment]
@L-M-Sherlock, I create this function
``` rust
fn extract_first_long_term_reviews<'a>(
    items: impl IntoIterator<Item = &'a FSRSItem>,
) -> Vec<FSRSItem> {
    items
        .into_iter()
        .filter_map(|i| {
            let a = i
                .reviews
                .iter()
                .take_while_inclusive(|r| r.delta_t < 1)
                .cloned() // Close my open PR to make it .copied() and 1.5x faster
                .collect_vec();
            if a.last()?.delta_t < 1 || a.len() == i.reviews.len() {
                return None;
            }
            Some(FSRSItem { reviews: a })
        })
        .collect()
}
```
It uses Itertools and runs in 1.5s for 10'000'000 items.
